core/com.b2international.snowowl.datastore.server/src/com/b2international/snowowl/datastore/server/internal/CDOBasedRepository.java

```diff
@@ -348,5 +321,102 @@ public void handleCommitInfo(CDOCommitInfo commitInfo) {
 			notification.publish(events());
         }
 	}
+
+	/**
+	 * Verifies if the repository is in a consistent state, and sets the {@link #state repositoryState} accordingly
+	 */
+	@Override
+	public void updateHealth() {
+		Health newHealth = calculateHealth();
+		if (health != newHealth) {
+			LOG.debug("Repository state for {} is changing from {} to {}", getCdoRepository().getRepositoryName(), health, newHealth);
+			health = newHealth;
+		} else {
+			LOG.debug("Repository state for {} is remaining the same: {}", getCdoRepository().getRepositoryName(), health);
+		}
+	}
 	
+	@Override
+	public Health getHealth() {
+		return health;
+	}
+	
+	public Health calculateHealth() {
+		IBranchPath branchPath = BranchPathUtils.createMainPath();
+		List<CDOCommitInfo> cdoCommitInfos = getCDOCommitInfos(branchPath);
+		CommitInfos indexCommitInfos = getIndexCommitInfos(branchPath);
+		
+		boolean emptyDb = cdoCommitInfos.isEmpty();
+		boolean emptyIndex = indexCommitInfos.getItems().isEmpty();
+
+		if (emptyDb && emptyIndex) {
+			return Health.GREEN; // empty dataset
+		}
+		
+		if (emptyDb ^ emptyIndex) {
+			LOG.error("{} is in inconsistent state. CDO {} but INDEX {}", getCdoRepository().getRepositoryName(), contentMessage(emptyDb), contentMessage(emptyIndex));
+			return Health.RED; // either CDO or index was deleted but not the other.
+		}
+		
+		boolean validContent = !hasInvalidCDOTimeStamps(cdoCommitInfos, getLast(indexCommitInfos).getTimeStamp(), branchPath);
+		
+		return validContent == true ? Health.GREEN : Health.RED;
+	}
+
+	private String contentMessage(boolean empty) {
+		return empty ? "IS EMPTY" : "HAS CONTENT";
+	}
+
+	private boolean hasInvalidCDOTimeStamps(List<CDOCommitInfo> cdoCommitInfos, long lastIndexCommitTimestamp, IBranchPath branch) {
```

Comment:
If you use only the `path` part of an `IBranchPath`, then it is recommended to use a String to represent your path.